client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java

```diff
@@ -177,8 +180,20 @@ private ChannelPool newPool(SocketAddress address) {
       if (maxConnections < 0) {
          maxConnections = Integer.MAX_VALUE;
       }
-      ChannelInitializer channelInitializer = new ChannelInitializer(bootstrap, address, operationsFactory, configuration, this);
-      bootstrap.handler(channelInitializer);
+
+      SslHandlerFactory sslHandlerFactoryForChannelInitializer = configuration.security().ssl().enabled() && !configuration.isSinglePort()?
+            alloc -> SslHandlerHelper.createSslHandler(configuration, alloc) : null;
+      SslHandlerFactory sslHandlerFactoryForSinglePortAdapter = configuration.security().ssl().enabled() && configuration.isSinglePort() ?
```

Comment:
This code can be improved, with a single check of `isSinglePort` and do all you need inside there. With the current code you're checking this twice, in this lines and then shortly below, making it more confusing to read.

Reply:
Unfortunately those objects are tightly coupled and decomposing them as you proposed won't be easy. Here's why...

The `ChannelInitializer` requires an `SslHandlerFactory`, so it means it needs to be initialized *after* it. The `CustomProtocolUpgradeCodec` needs a Netty pipeline that needs to plugged in after a successful upgrade. So again, `CustomProtocolUpgradeCodec` comes after `ChannelInitializer` (since it creates the pipeline). The `SinglePortAdapter` needs am upgrade codec (long story short - this is due to Netty's HTTP/1.1 design), so it comes after `CustomProtocolUpgradeCodec` and an `SslHandlerFactory` (because it can operate with or without SSL). Finally, the `ChannelPool` requires `ChannelInitializer` and vice versa. So here's the creation order:

1. `SslHandlerFactory`
2. `ChannelInitializer`
3. `CustomProtocolUpgradeCodec`
4. `SinglePortAdapter`
5. `ChannelPool`

I can swap if/else statements but I need to preserve exactly this order. That means I will be either branching by `configuration.isSinglePort()` or `configuration.security().ssl().enabled()`. This way or another the amount of if/else statements will remain the same (I've just checked it in my IDE). 

If you have any other idea how to refactor it, please paste me a code snippet (but please check it in your IDE before that....this one looks easy only on paper :D).

Reply:
I think Galder meant something like
```java
SslHandlerFactory sslHandlerFactoryForChannelInitializer = null;
SslHandlerFactory sslHandlerFactoryForSinglePortAdapter = null;
if (configuration.security().ssl().enabled()) {
   if (configuration.isSinglePort()) {
      sslHandlerFactoryForSinglePortAdapter = ...
   } else {
      sslHandlerFactoryForChannelInitializer = ...
   }
}

Reply:
Ok, implemented.